### PR TITLE
Add convenience functions for overriding ABMs, LBMs and entities

### LINKED ISSUE #6909: Feature request: __builtin:item on_step callback
The "dropped item"  entity '__builtin:item' should have an callback in on_step to allow game-specific behaviors like other physics, environment manipulations, light, knock-back, and many others using mods. The best way is a function that is executed before the builtin logic and if the function returns false the builtin logic is skipped. Something like minetest.register_item_entity_on_step(func(obj)) ?

### PR DESCRIPTION
**Pull request by sfence**
- Extend Lua API for mod and game creators.
- It adds override Lua functions for ABM, LBM, and entities.
- This should fix #6909 at least.
- This also adds some checks to generate an error message when some mod tries to call some functions like `register_node`, `register_lbm`, etc after mod loading is finished.

## To do

This PR is a Ready for Review 

- [x] Get feedback.
- [x] Apply feedback.
- [x] Add some tests, probably to devtest unittests.
- [x] Check if biome definition and/or similar definitions can be overridden as well ( #9161 )

## How to test

Run detest game. `minetest.override_entity` is tested in unittests. Test for `minetest.override_abm` and `minetest.override_lbm` can be done by placing nodes `testabms:override` and `testlbms:onload_change` from mods testabms and testlbms.

### CODE REVIEW

**Comment by Zughy** [comment]
(only entities are in the roadmap, but I guess it won't hurt :woman_shrugging: ; unsubscribing)

**Comment by grorp** [comment]
- Adding error messages about calling registration functions after load time could go into a dedicated bugfix PR.

- At least entity definitions are definitely already overridable by modifying `minetest.registered_entities`:
    
    https://github.com/minetest/minetest/blob/32e492837cbf286aeb91b4b63ecf3c890c71a1bc/doc/lua_api.md?plain=1#L6992-L6996
   
   This even works after load time AFAIK. So this PR only makes it more convenient.

- Shouldn't overriding ABMs and LBMs by modifying `minetest.registered_abms`/`minetest.registered_lbms` at load time already work as well?

Of course, that doesn't make this PR invalid. It's more convenient to have API functions for this purpose.

**Comment by kromka-chleba** [comment]
>It adds override Lua functions for ABM, LBM, and entities.

I guess this resolves my issue? https://github.com/minetest/minetest/issues/13112
TL;DR Will this let me unregister and change ABMs/LBMs?

**Comment by sfence** [comment]
> #13112

This only allows a change of abm/lbm at loading time. After mods are loaded, the Lua tables `registered_abm` and `registered_lbm` are loaded by C++ and converted to C++ class/structures which are used by the engine to proceed abm and lbm calls.

Supporting dynamic changing of lbm/abm during the game running will be more difficult to implement.

**Comment by sfence** [comment]
> * Adding error messages about calling registration functions after load time could go into a dedicated bugfix PR.

If this PR is closed without merging, I can create a specific PR for this. But because it takes time to merge PR, I would like to keep it here, instead of creating a separate PR and setting this PR to be blocked/waiting for it to be merged. 
> * At least entity definitions are definitely already overridable by modifying `minetest.registered_entities`:
>   https://github.com/minetest/minetest/blob/32e492837cbf286aeb91b4b63ecf3c890c71a1bc/doc/lua_api.md?plain=1#L6992-L6996

You are right. But in this way, with added error message it will be a more foolproof/error-proof way to override. This is too reason why do not create separate bugfix PR, I think.
>   
>   This even works after load time AFAIK. So this PR only makes it more convenient.

Yes, but at least adding `override_entity` is in the roadmap. If actual solution is enough, Issue #6909 should in the closed state.
> * Shouldn't overriding ABMs and LBMs by modifying `minetest.registered_abms`/`minetest.registered_lbms` at load time already work as well?

Yes, it is possible. But you have to go through `register_abms` or `registered_lbms` tables manually in every situation where you want to override abm or lbm.
> Of course, that doesn't make this PR invalid. It's more convenient to have API functions for this purpose.

**Comment by sfence** [comment]
Overriding of biomes will require changes and method adding in C++, so I will not implement it here. It should be a separate PR.

**Comment by sfence** [comment]
I have added the possibility of testing override functionality.

So, this is now prepared for review.

**Comment by lhofhansl** [comment]
I like the idea. Nice tests!
I find myself not in a position to really review it. It looks right, and nothing obviously bad.

Once question, could one also do this with a helper mod that implements `override_abm`? It seems all this does it looking through the list of registered ABMs/LBMs and then modifies them.

**Comment by sfence** [comment]
> I like the idea. Nice tests! I find myself not in a position to really review it. It looks right, and nothing obviously bad.
> 
> Once question, could one also do this with a helper mod that implements `override_abm`? It seems all this does it looking through the list of registered ABMs/LBMs and then modifies them.

Yes, all of this can be done by external mod with an exception for checking the `is_mods_loaded` field.
I have added and documented an optional new field `name` for ABM. It is checked in `register_abm` calls, if is provided, and allows ABM to be overridden.
It cannot be a required field because of backward compatibility.

But, my opinion is that It will be better to have this as part of the official modding API.

**Comment by Zughy** [comment]
@sfence more rebase needed

**Comment by Zughy** [comment]
@sfence rebase needed

**Comment by farooqkz** [comment]
I guess an ABM can be removed by setting `action = nil`, right? We could make good use of this in JMA CTF. Will be happy to have it merged :)

**Comment by grorp** [comment]
> I guess an ABM can be removed by setting `action = nil`, right? We could make good use of this in JMA CTF. Will be happy to have it merged :)

This PR only allows changing ABMs during load time, not runtime, is that what you want? If it is, you should be able to achieve the same thing by manually modfiying `core.registered_abms` for now.

https://github.com/luanti-org/luanti/blob/eef2097e080c7000a05e4f6661dc260971910bc5/builtin/game/register.lua#L110-L114

If you want changes at runtime, there is #15992.

**Comment by farooqkz** [comment]
> > I guess an ABM can be removed by setting `action = nil`, right? We could make good use of this in JMA CTF. Will be happy to have it merged :)
> 
> This PR only allows changing ABMs during load time, not runtime, is that what you want? If it is, you should be able to achieve the same thing by manually modfiying `core.registered_abms` for now.
> 
> https://github.com/luanti-org/luanti/blob/eef2097e080c7000a05e4f6661dc260971910bc5/builtin/game/register.lua#L110-L114
> 
> If you want changes at runtime, there is #15992.

hmm what do you mean by "load" time? when exactly? that PR is closed unfortunately :(

**Comment by grorp** [comment]
> hmm what do you mean by "load" time? when exactly?

in the case of this PR, the last point in time at which ABMs can be changed is in a `core.register_on_mods_loaded` callback. this is e.g. before any server steps happen or players can join
